File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.js
/**
 * @param {number[]} bloomDay
 * @param {number} m
 * @param {number} k
 * @return {number}
 */
var minDays = function (bloomDay, m, k) {
    function calculate(mid, bloomDay, m, k) {
        let count = 0;
        let bouquets = 0;
        for (let i = 0; i < bloomDay.length; i++) {
            if (bloomDay[i] <= mid) {
                count++;
            }
            else {
                bouquets = bouquets + Math.floor(count / k);
                count = 0;
            }

        }
        bouquets = bouquets + Math.floor(count / k);
        return bouquets;

    }

    let low = 1;
    let high = Math.max(...bloomDay)
    let res=-1;
    while (low <= high) {
        let mid = Math.floor((low + high) / 2);
        if (calculate(mid, bloomDay, m, k) >= m) {
            res = mid;
            high = mid - 1;
        }
        else {
            low = mid + 1;
        }
    }
    return res;
};

File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    bool helper(vector<int>& bloomDay, int m, int k, int mid) {
        int current = 0;
        int count = 0;
        for (int i = 0; i < bloomDay.size(); i++) {
            if (bloomDay[i] <= mid) {
                current++;
            } else {
                count = count + (current / k);
                current = 0;
            }
        }
        count = count + (current / k);
        if(count>=m) return true;
        return false;
    }
    int minDays(vector<int>& bloomDay, int m, int k) {
        int low = 1;
        int high = bloomDay[0];
        for(int i=0;i<bloomDay.size();i++){
            high = max(high, bloomDay[i]);
        }
        cout<<helper(bloomDay, m, k, high);
        int ans =-1;
        while(low<=high) {
            int mid = (low + high) / 2;
           
            if (helper(bloomDay, m, k, mid)) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return ans;
    }
};